ejb3/src/main/java/org/jboss/as/ejb3/suspend/EJBSuspendHandlerService.java

```diff
@@ -0,0 +1,338 @@
+
+    /**
+     * Sets a new value for {@code gracefulTxnShutdown}.
+     *
+     * @param gracefulTxnShutdown new value of the model attribute
+     */
+    public void enableGracefulTxnShutdown(boolean gracefulTxnShutdown) {
+        this.gracefulTxnShutdown = gracefulTxnShutdown;
+    }
+
+    /**
+     * Returns suspend controller injected value.
+     *
+     * @return suspend controller injected value
+     */
+    public InjectedValue<SuspendController> getSuspendControllerInjectedValue() {
+        return suspendControllerInjectedValue;
+    }
+
+    /**
+     * Returns local transaction context injected value.
+     *
+     * @return local transaction context injected value
+     */
+    public InjectedValue<LocalTransactionContext> getLocalTransactionContextInjectedValue() {
+        return localTransactionContextInjectedValue;
+    }
+
+    /**
+     * Returns service value.
+     */
+    @Override public EJBSuspendHandlerService getValue() {
+        return this;
+    }
+
+    /**
+     * Starts the service. Registers server activity, sets transaction listener on local transaction context, and creates and
+     * installs deployment controller service.
+     *
+     * @param context start context
+     */
+    public void start(StartContext context) {
+        final SuspendController suspendController = suspendControllerInjectedValue.getValue();
+        suspendController.registerActivity(this);
+        final LocalTransactionContext localTransactionContext = localTransactionContextInjectedValue.getValue();
+        localTransactionContext.registerCreationListener(this);
+        deploymentController = context.getChildTarget().addService(DEPLOYMENT_CONTROLLER, new AbstractService<Void>() {})
+                .setInitialMode(ServiceController.Mode.ACTIVE).install();
+
+    }
+
+    /**
+     * Stops the service. Unregisters service activity and clears transaction listener.
+     * @param context stop context
+     */
+    public void stop(StopContext context) {
+        final SuspendController suspendController = suspendControllerInjectedValue.getValue();
+        suspendController.unRegisterActivity(this);
+        final LocalTransactionContext localTransactionContext = localTransactionContextInjectedValue.getValue();
+        localTransactionContext.removeCreationListener(this);
+    }
+
+    /**
+     * Pre suspend. Do nothing.
+     * @param listener callback listener
+     */
+    @Override public void preSuspend(ServerActivityCallback listener) {
+        listener.done();
+    }
+
+    /**
+     * Notifies local transaction context that server is suspended, and only completes suspension if
+     * there are no active invocations nor transactions.
+     *
+     * @param listener callback listener
+     */
+    @Override public void suspended(ServerActivityCallback listener) {
+        this.suspended = true;
+        listenerUpdater.set(this, listener);
+        localTransactionContextInjectedValue.getValue().suspendRequests();
```

Comment:
This is wrong, as it means that existing requests may not be able to complete as they cannot create a transaction.